primap2/_setters.py

```diff
@@ -0,0 +1,516 @@
+            if value_dims is not None:
+                raise ValueError("value_dims given, but value is already a DataArray.")
+
+            # conform value to given dim: key
+            if dim not in value.dims:
+                if dim in value.coords:
+                    value = value.reset_coords(dim, drop=True)
+                value = value.expand_dims({dim: key})  # type: ignore
+            else:
+                value = value.loc[{dim: key}]
+            value, expanded = xr.broadcast(value, self._da)
+            value.attrs = self._da.attrs
+            value.name = self._da.name
+            sel = expanded.loc[{dim: key}]
+
+        else:
+            new_index = list(self._da[dim].values)
+            for item in key:
+                if item not in new_index:
+                    new_index.append(item)
+            new_index = np.array(new_index, dtype=self._da[dim].dtype)
+            expanded = self._da.reindex({dim: new_index}, copy=False)
+            sel = expanded.loc[{dim: key}]
+
+            if value_dims is None:
+                value_dims = []
+                for i, idim in enumerate(sel.dims):
+                    if sel.shape[i] > 1:
+                        value_dims.append(idim)
+                if len(value_dims) != len(value.shape):
+                    raise ValueError(
+                        "Could not automatically determine value dimensions, please"
+                        " use the value_dims parameter."
+                    )
+            value = xr.DataArray(
+                value,
+                coords=[(idim, sel[idim]) for idim in value_dims],
+                name=self._da.name,
+                attrs=self._da.attrs,
+            ).broadcast_like(sel)
+
+        if existing == "fillna_empty":
+            if sel.count().item() > 0:
+                already_existing = set(self._da[dim].values).intersection(set(key))
+                raise ValueError(
+                    f"Values {already_existing!r} for {dim!r} already exist and contain"
+                    f" data. Use existing='overwrite' or 'fillna' to avoid this error."
+                )
+            existing = "fillna"
+
+        if existing == "overwrite":
+            return value.combine_first(expanded)
+        elif existing == "fillna":
+            return expanded.combine_first(value)
+        else:
+            raise ValueError(
+                "If given, 'existing' must specify one of 'error', 'overwrite', "
+                f"'fillna_empty', or 'fillna', not {existing!r}."
+            )
+
+
+class DatasetSettersAccessor(_accessor_base.BaseDatasetAccessor):
+    @staticmethod
+    def _set_apply(
+        da: xr.DataArray,
+        dim: typing.Hashable,
+        key: typing.Any,
+        value: xr.Dataset,
+        existing: str,
+    ) -> xr.DataArray:
+        if dim not in da.dims:
+            return da
+        return da.pr.set(dim=dim, key=key, value=value[da.name], existing=existing)
+
+    def set(
```

Comment:
How can set be limited to some or just one variable(s)?

Reply:
Given a Dataset with multiple variables, limiting set to e.g. "CO2" would be done like this:
```python
ds["CO2"] = ds["CO2"].pr.set(…)
```
i.e. use the variable's set method instead of the Datasets set method.